### PR TITLE
feat(provider/cf): add deploy, server group, and load balancer operations

### PR DESCRIPTION
**Pull request by jkschneider**
Adds all basic operations to the Cloud Foundry implementation. We'll submit the create/destroy service (Open Service Broker) code separately.

### CODE REVIEW

File: clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/cache/CacheRepository.java [hunk 1]
```diff
@@ -120,8 +123,9 @@ private CloudFoundryServerGroup serverGroupFromCacheData(CacheData serverGroupDa
 
   private CloudFoundryLoadBalancer loadBalancerFromCacheData(CacheData lbData, Detail detail) {
     CloudFoundryLoadBalancer loadBalancer = objectMapper.convertValue(lbData.getAttributes().get("resource"), CloudFoundryLoadBalancer.class);
-    if (detail.equals(Detail.NONE))
+    if (detail.equals(Detail.NONE)) {
```
Comment (lwander) [inline clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/cache/CacheRepository.java]: :+1:

File: clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/converters/DeployCloudFoundryServerGroupAtomicOperationConverter.java [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+  public DeployCloudFoundryServerGroupDescription convertDescription(Map input) {
+    DeployCloudFoundryServerGroupDescription converted = getObjectMapper().convertValue(input, DeployCloudFoundryServerGroupDescription.class);
+    CloudFoundryCredentials credentials = getCredentialsObject(input.get("credentials").toString());
+    converted.setClient(credentials.getClient());
+    converted.setAccountName(credentials.getName());
+
+    converted.setSpace(findSpace(converted.getRegion(), converted.getClient())
+      .orElseThrow(() -> new IllegalArgumentException("Unable to find space '" + converted.getRegion() + "'.")));
+
+    Map artifactSource = (Map) input.get("artifactSource");
```
Comment (lwander) [inline clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/converters/DeployCloudFoundryServerGroupAtomicOperationConverter.java]: Are these intended to be the canonical spinnaker artifacts?

File: clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/converters/DeployCloudFoundryServerGroupAtomicOperationConverter.java [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+    if ("artifact".equals(artifactSource.get("type"))) {
```
Comment (lwander) [inline clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/converters/DeployCloudFoundryServerGroupAtomicOperationConverter.java]: What does an artifact with type "artifact" mean?

File: clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/converters/DeployCloudFoundryServerGroupAtomicOperationConverter.java [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+      converted.setArtifact(artifact);
+    }
+
+    Map manifest = (Map) input.get("manifest");
+    if ("direct".equals(manifest.get("type"))) {
+      converted.setApplicationAttributes(getObjectMapper().convertValue(manifest,
+        DeployCloudFoundryServerGroupDescription.ApplicationAttributes.class));
+    } else if ("artifact".equals(manifest.get("type"))) {
+      Artifact manifestArtifact = convertToArtifact(manifest.get("account").toString(), manifest.get("reference").toString());
+      ArtifactCredentials manifestArtifactCredentials = credentialsRepository.getAllCredentials().stream()
```
Comment (lwander) [inline clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/converters/DeployCloudFoundryServerGroupAtomicOperationConverter.java]: I think this logic should live in the artifacts module in clouddriver -- finding the credentials & downloading artifacts is a common-enough paradigm to share the code.

File: clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/ops/CloudFoundryOperationUtils.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+import com.netflix.spinnaker.clouddriver.cloudfoundry.client.model.v3.ProcessStats;
+
+class CloudFoundryOperationUtils {
+  static String describeProcessState(ProcessStats.State state) {
+    switch(state) {
+      case STARTING:
+        return "is still starting";
+      case CRASHED:
+        return "crashed";
```
Comment (lwander) [inline clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/ops/CloudFoundryOperationUtils.java]: maybe "has crashed" for consistency?

File: clouddriver-cloudfoundry/src/test/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/ops/AbstractCloudFoundryAtomicOperationTest.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
+/*
```
Comment (lwander) [inline clouddriver-cloudfoundry/src/test/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/ops/AbstractCloudFoundryAtomicOperationTest.java]: FWIW it's totally fine to write tests in groovy if you find it easier -- not necessary either way though.

File: clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/converters/DeployCloudFoundryServerGroupAtomicOperationConverter.java (hunk 2 above)
Comment (jkschneider) [reply clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/converters/DeployCloudFoundryServerGroupAtomicOperationConverter.java]: No. This is being replaced by the expected artifacts improvement in the end.

File: clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/converters/DeployCloudFoundryServerGroupAtomicOperationConverter.java (hunk 3 above)
Comment (jkschneider) [reply clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/converters/DeployCloudFoundryServerGroupAtomicOperationConverter.java]: This is being replaced by the expected artifacts improvement in the end.

File: clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/converters/DeployCloudFoundryServerGroupAtomicOperationConverter.java (hunk 4 above)
Comment (jkschneider) [reply clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/converters/DeployCloudFoundryServerGroupAtomicOperationConverter.java]: This is being replaced by the expected artifacts improvement in the end.

File: clouddriver-cloudfoundry/src/test/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/ops/AbstractCloudFoundryAtomicOperationTest.java (hunk 6 above)
Comment (jkschneider) [reply clouddriver-cloudfoundry/src/test/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/ops/AbstractCloudFoundryAtomicOperationTest.java]: At this point I really don't like Groovy much. Java is fine.

File: clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/ops/CloudFoundryOperationUtils.java (hunk 5 above)
Comment (jkschneider) [reply clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/ops/CloudFoundryOperationUtils.java]: The text of a message coming from this reads: "Failed to start 'demo' which instead crashed".

**Pull request merged**